### DISCUSSION TITLE
Provider decoded key - but now what?

### DISCUSSION THREAD

**Discussion by commel** [openssl/openssl#24500] [Q&A] [answered]
My provider decoded its key successfully. How do you pass this around for openssl to use this like for creating a certificate? When trying to attempt this, I keep getting messages about how the key cannot be read. But it already did so in the decoder? Is this an export problem?

```
Could not read key from example-priv.pem
808B7EA96B7A0000:error:1608010C:STORE routines:ossl_store_handle_load_result:unsupported:../crypto/store/store_result.c:151:
```

I am already calling the callback function and I've prepared the key in the OSSL PARAMS-format like this:
```
    OSSL_PARAM params[4];
    int object_type = OSSL_OBJECT_PKEY;

    params[0] = OSSL_PARAM_construct_int(OSSL_OBJECT_PARAM_TYPE, &object_type);
    params[1] = OSSL_PARAM_construct_utf8_string(OSSL_OBJECT_PARAM_DATA_TYPE, (char*)"privkey", 0);
    params[2] = OSSL_PARAM_construct_octet_string(OSSL_OBJECT_PARAM_DATA, (void*)oct->data, oct->length);    
    params[3] = OSSL_PARAM_construct_end();
```
But nothing happens - or changes - even if I change the parameter types.

According to the decoder documentation (https://www.openssl.org/docs/man3.3/man7/provider-decoder.html) i should provide a provider native export:

```
The last decoding step in the decoding chain is usually supposed to create a provider-native object referenced by an object reference. 
To import that object into a different provider the OSSL_FUNC_decoder_export_object() can be called as the 
final step of the decoding process.
```
The given format above works fine in the keymanagement, so I figured this would be it?

**Comment by levitte** [comment]
This looks strange:

``` C
    params[1] = OSSL_PARAM_construct_utf8_string(OSSL_OBJECT_PARAM_DATA_TYPE, (char*)"privkey", 0);
```

[provider-object(7)](https://www.openssl.org/docs/man3.3/man7/provider-object.html) says this about that parameter:

> "data-type" (**OSSL_OBJECT_PARAM_DATA_TYPE**) <UTF8 string>
> 
>   The specific type of the object content. Legitimate values depend on the object type; if it is **OSSL_OBJECT_PKEY**, the data type is expected to be a key type suitable for fetching a [provider-keymgmt(7)](https://www.openssl.org/docs/man3.3/man7/provider-keymgmt.html) that can handle the data.

So unless you have a keytype (keymgmt name) called "privkey", that won't be useful.
(for example, a decoded RSA key would have the data type "rsa").

Although, may I ask, what key type are you actually handling, and exactly what are the contents of `oct->data`?

**Reply by commel** [reply]
Thanks for getting back on this! The key is a ED25519 under a custom OID, which my encoder successfully wrote into a key file:
```
openssl asn1parse -in example-priv.pem 
    0:d=0  hl=2 l=  46 cons: SEQUENCE          
    2:d=1  hl=2 l=   1 prim: INTEGER           :00
    5:d=1  hl=2 l=   5 cons: SEQUENCE          
    7:d=2  hl=2 l=   3 prim: OBJECT            :1.2.3.4
   12:d=1  hl=2 l=  34 prim: OCTET STRING      [HEX DUMP]:0420068A9857A95BF1E1EEFD7B1D0D6D9CEAB4F44D25C6B827FA2E810FCFE29FDF07
```
I see calls for KEYMGMT after DECODE calls. The key management import handles the "privkey" PARAM, but my specific implementation does not seem to be called at all after decode.

**Reply by levitte** [reply]
Ah, that's the same thing you talk about in #24447, I take it.

So, judging on that, you have created a new key type ("xyz" with OID 1.2.3.4), which looks a behaves like an ED25519, but from a key type perspective is very much its own type.

Your decoders (#24447) very clearly say that are to decode an "xyz" key from PEM or DER, respectively.  So in your example above, this change should be made:

``` diff
-    params[1] = OSSL_PARAM_construct_utf8_string(OSSL_OBJECT_PARAM_DATA_TYPE, (char*)"privkey", 0);
+    params[1] = OSSL_PARAM_construct_utf8_string(OSSL_OBJECT_PARAM_DATA_TYPE, (char*)"xyz", 0);
```

So you end up needing a keymgmt for "xyz".

But, that's not all, your "xyz" decoders must actually be able to decode the key data into data that the "xyz" keymgmt can use directly, and that data needs to be passed by reference (`OSSL_OBJECT_PARAM_REFERENCE` param) rather than by value (`OSSL_OBJECT_PARAM_DATA`).
Whatever you pass by references here will be passed to the "xyz" keymgmt's load function (dispatch function index `OSSL_FUNC_KEYMGMT_LOAD`), and if this can be used quite creatively.

So, your "xyz" decoders need to produce something that can be passed to your "xyz" keymgmt load function.  When that works right, libcrypto's decoder machinery will be able to wrap that keymgmt instance in an EVP_PKEY.

**Reply by levitte** [reply]
I hope what I said makes some sort of sense.

**Reply by commel** [reply]
Oh great insight! I wasnt aware I had to use the load-function. What am I supposed to return here? Internally I work with an EVP_PKEY, should I create it from the raw key in the LOAD function?

**Reply by levitte** [reply, marked as answer]
Actually, now that I think of it, you can pass whatever you want.  Essentially, the decoding to a keytype becomes a dialogue between whatever your decoder (through `OSSL_OBJECT_PARAM_REFERENCE`) and the corresponding keymgmt's load function.  So, you *can* pass a DER blob that way, *as long as* the corresponding keymgmt's load function treats its input reference as a DER blob and does the last bit of decoding into the internal EVP_PKEY you use.

The choice we made in OpenSSL's providers was to let the decoders decode into the actual internal key structure (for you, it would be an internal EVP_PKEY) and let the keymgmt load function handle it, see here:

https://github.com/openssl/openssl/blob/05faa4ffee7f20fcee129f77d153f2dcc609bdc8/providers/implementations/encode_decode/decode_der2key.c#L309-L326

and here for an example load function:

https://github.com/openssl/openssl/blob/05faa4ffee7f20fcee129f77d153f2dcc609bdc8/providers/implementations/keymgmt/ecx_kmgmt.c#L830-L842

Makes sense?

**Reply by levitte** [reply]
(do note that our der2key decoder handles all our key types in the same code block...  you won't have to do something near as complex)

**Reply by commel** [reply]
Thanks for clarifying it! This comes pretty close to what I've had. I've adapted the name to "ed25519-xyz". I've referenced the LOAD-function in the dispatch for decode, but somehow this is not called at all. 

The decode dispatch table is defined as such:
```
     {"ed25519-xyz:1.2.3.4, "provider=xyz,input=pem", dispatch_decoder_ed25519_private_der, PROV_DESCS_ED25519}
```
which should match the key's definition, right?

**Reply by commel** [reply]
I've split up the keymgmt into two parts (gen and load) which both had the same identifier, so probably only the first one was loaded. I now get to the encoder part of the public key, which is a huge progress. Thank you @levitte!
